internal/uvm/vnuma.go

```diff
@@ -0,0 +1,184 @@
+	}
+	return numa, nil, validate(numa)
+}
+
+const (
+	WildcardPhysicalNodeNumber = 0xFF
+	NumaTopologyNodeCountMax   = 64
+	NumaChildNodeCountMax      = 64
+)
+
+// Validate validates self-contained fields within the given NUMA settings.
+//
+// TODO (maksiman): Check if we need to add compute-less node validation. For now, assume that it's supported.
+func validate(n *hcsschema.Numa) error {
+	if len(n.Settings) == 0 {
+		// Nothing to validate
+		return nil
+	}
+
+	var virtualNodeSet = make(map[uint32]struct{})
+	var virtualSocketSet = make(map[uint32]struct{})
+	var totalVPCount uint32
+	var totalMemInMb uint64
+	var highestVNodeNumber uint32
+	var highestVSocketNumber uint32
+
+	hasWildcardPhysicalNode := n.Settings[0].PhysicalNodeNumber == WildcardPhysicalNodeNumber
+
+	for _, topology := range n.Settings {
+		if topology.VirtualNodeNumber > NumaChildNodeCountMax {
+			return fmt.Errorf("vNUMA virtual node number %d exceeds maximum allowed value %d", topology.VirtualNodeNumber, NumaChildNodeCountMax)
+		}
+		if topology.PhysicalNodeNumber != WildcardPhysicalNodeNumber && topology.PhysicalNodeNumber >= NumaTopologyNodeCountMax {
+			return fmt.Errorf("vNUMA physical node number %d exceeds maximum allowed value %d", topology.PhysicalNodeNumber, NumaTopologyNodeCountMax)
+		}
+		if hasWildcardPhysicalNode != (topology.PhysicalNodeNumber == WildcardPhysicalNodeNumber) {
+			return fmt.Errorf("vNUMA has a mix of wildcard (%d) and non-wildcard (%d) physical node numbers", WildcardPhysicalNodeNumber, topology.PhysicalNodeNumber)
+		}
+
+		if topology.CountOfMemoryBlocks == 0 {
+			return fmt.Errorf("vNUMA nodes with no memory are not allowed")
+		}
+
+		totalVPCount += topology.CountOfProcessors
+		totalMemInMb += topology.CountOfMemoryBlocks
+
+		if _, ok := virtualNodeSet[topology.VirtualNodeNumber]; ok {
+			return fmt.Errorf("vNUMA virtual node number %d is duplicated", topology.VirtualNodeNumber)
+		}
+		virtualNodeSet[topology.VirtualNodeNumber] = struct{}{}
+
+		if topology.MemoryBackingType != hcsschema.MemoryBackingType_PHYSICAL {
+			return fmt.Errorf("vNUMA memory backing type %s is invalid", topology.MemoryBackingType)
+		}
+
+		if highestVNodeNumber < topology.VirtualNodeNumber {
+			highestVNodeNumber = topology.VirtualNodeNumber
+		}
+		if highestVSocketNumber < topology.VirtualSocketNumber {
+			highestVSocketNumber = topology.VirtualSocketNumber
+		}
+
+		virtualSocketSet[topology.VirtualSocketNumber] = struct{}{}
+	}
+
+	// Either both total memory and processor count should be zero or both should be non-zero
+	if (totalMemInMb == 0) != (totalVPCount == 0) {
+		return fmt.Errorf("partial resource allocation is not allowed")
+	}
+
+	// At least
+	if totalMemInMb == 0 && hasWildcardPhysicalNode {
+		return fmt.Errorf("completely empty topology is not allowed")
+	}
+
+	if len(virtualNodeSet) != int(highestVNodeNumber+1) {
+		return fmt.Errorf("holes in vNUMA node numbers are not allowed")
```

Comment:
what does this mean? Like we couldn't use nodes 0 and 2?

Reply:
Since vCPU node number is generated implicitly, i think this check is not required. 
May be this check was borrowed from WMI API where vCPU node number can be explicitly specified by the user.
@anmaxvl : Is it ok to remove the check for VirtualNodeNumber and VirtualSocketNumber?

Reply:
yeah, that's fair.